### PR TITLE
Write tests for CSSOM Interfaces

### PR DESCRIPTION
**Pull request by canova**
ToCss implementation was wrong about Keyframe percentage values, it was writing values between 0-1. I had to fix it. Wrote some tests about CSSKeyframesRule, CSSNamespaceRule, CSSRuleList, CSSStyleSheet, StyleSheetList interfaces. 
CSSFontFaceRule and CSSViewportRule looks like not implemented yet. I didn't write one for them.
Also name attribute in CSSKeyframesRule isn't implemented yet. Is there any complication about it? If not, I can implement it.

r? @Manishearth 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] There are tests for these changes OR

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14464)

### CODE REVIEW

**Comment by Ms2ger** [comment]
Please also fix
```
/// A number from 1 to 100, indicating the percentage of the animation where
/// this keyframe should run.
```

**Comment by canaltinova** [comment]
Fixed this and added ToCss for KeyframePercentage. Thanks!

**Comment by Ms2ger** [comment]
Thanks! I'm going to leave the tests to someone else, but the code changes make sense to me.

File: tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        body { width: 50%; }
+        #foo { height: 100px; }
+    </style>
+
+    <script>
+    test(function () {
+        var styleSheet = document.styleSheets[0];
+        assert_equals(styleSheet.cssRules.length, 2, "CSSStyleSheet cssRules attribute");
+        assert_equals(styleSheet.cssRules[0].cssText, "body { width: 50%; }", "CSSStyleSheet cssRules attribute function");
+        assert_equals(styleSheet.cssRules[1].cssText, "#foo { height: 100px; }", "CSSStyleSheet cssRules attribute function");
```
Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: attribute, not attribute function

File: tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharnessreport.js"></script>
+    <style>
+        body { width: 50%; }
+        #foo { height: 100px; }
+    </style>
+
+    <script>
+    test(function () {
+        var styleSheet = document.styleSheets[0];
```
Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: Ensure that it's equal to `document.getElementById("my-stylesheet").style` (giving the `<style>` element an id above)

File: tests/wpt/web-platform-tests/cssom/StyleSheetList.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    test(function () {
+        var styleSheets = document.styleSheets;
+        assert_equals(styleSheets.length, 2, "StyleSheetList length attribute");
+        assert_equals(styleSheets[0].cssRules.length, 1, "StyleSheetList indexed getter length attribute");
+        assert_equals(styleSheets[1].cssRules.length, 2, "StyleSheetList indexed getter length attribute");
+        assert_equals(styleSheets[2], undefined, "StyleSheetList indexed getter length attribute");
+        assert_equals(styleSheets.item(0).cssRules.length, 1, "StyleSheetList item function length attribute");
+        assert_equals(styleSheets.item(1).cssRules.length, 2, "StyleSheetList item function length attribute");
+        assert_equals(styleSheets.item(2), null, "StyleSheetList item function length attribute");
+    });
```
Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/StyleSheetList.html]: Ensure that `[SameObject]` holds here. If I run `styleSheets[0].randomProperty = 1`, inserting a new stylesheet into the document should not change `styleSheets[0].randomProperty`

**Review by Manishearth — APPROVED** [review]
Looks good, minor improvements.

File: tests/wpt/web-platform-tests/cssom/CSSKeyframesRule.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        assert_equals(find1.cssText, "50% { top: 100px; }", "CSSKeyframesRule findRule function");
+        var find2 = keyframe.findRule("0%");
+        assert_equals(find2.cssText, "0% { top: 0px; }", "CSSKeyframesRule findRule function");
+        var find3 = keyframe.findRule("70%");
+        assert_equals(find3, null, "CSSKeyframesRule findRule function");
+
+        keyframe.deleteRule("100%");
+        assert_equals(keyframe.cssRules.length, 2, "CSSKeyframesRule prefix attribute after deleteRule function");
+        assert_equals(keyframe.cssRules[0].cssText, "0% { top: 0px; }", "CSSKeyframesRule cssRule cssText attribute after deleteRule function");
+        assert_equals(keyframe.cssRules[1].cssText, "50% { top: 100px; }", "CSSKeyframesRule cssRule cssText attribute after deleteRule function");
```
Comment (canaltinova) [inline tests/wpt/web-platform-tests/cssom/CSSKeyframesRule.html]: https://drafts.csswg.org/css-animations/#interface-csskeyframesrule-appendrule says:
>The rule to be appended, expressed in the same syntax as one entry in the @keyframes rule. A valid rule is always appended e.g. even if its key(s) already exists.

So appending duplicate rule is valid. It's working as expected in servo currently. I'm adding a test for it

File: tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html (hunk 2 above)
Comment (canaltinova) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: `document.getElementById("my-stylesheet")` returns CSSStyleDeclaration and `document.styleSheets[0]` returns CSSStyleSheet.
`document.getElementById("my-stylesheet").style` returns this style element's associated style rules I guess.

**Comment by canaltinova** [comment]
Updated commits

File: tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html (hunk 2 above)
Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: `.style` also should return a CSSStyleSheet

Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: https://drafts.csswg.org/cssom/#linkstyle

Comment (canaltinova) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: Added a test for it, thanks.

File: tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<!doctype html>
+<html>
+<head>
+    <meta charset="utf-8">
+    <title>CSSOM - CSSStyleSheet interface</title>
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharnessreport.js"></script>
+    <style id="my-stylesheet">
+        body { width: 50%; }
```
Comment (canaltinova) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: Oops, added one for this, thanks

File: tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        assert_equals(styleSheet.cssRules[1].cssText, "#bar { margin: 10px; }", "CSSStyleSheet cssRules attribute after insertRule function");
+        assert_equals(styleSheet.cssRules[2].cssText, "#foo { height: 100px; }", "CSSStyleSheet cssRules attribute after insertRule function");
+
+        styleSheet.deleteRule(1);
+        assert_equals(styleSheet.cssRules.length, 2, "CSSStyleSheet cssRules attribute after deleteRule function");
+        assert_equals(styleSheet.cssRules[0].cssText, "body { width: 50%; }", "CSSStyleSheet cssRules attribute after deleteRule function");
+        assert_equals(styleSheet.cssRules[1].cssText, "#foo { height: 100px; }", "CSSStyleSheet cssRules attribute after deleteRule function");
+        assert_equals(styleSheet.cssRules[2], undefined, "CSSStyleSheet cssRules attribute after deleteRule function");
+
+        styleSheet.cssRules[0].randomProperty = 1;
```
Comment (Manishearth) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: I'd prefer if we set the random property (with different values) on `rules[0]` and `rules[1]` at the beginning of this test, and ensure that it stays after the insertion and deletion. The idea that the property should still stick to the css rule even after its index changes (so setting the property on `rules[1]`, running `insertRule`, and checking `rules[2]` will turn up the same property)

Comment (canaltinova) [inline tests/wpt/web-platform-tests/cssom/CSSStyleSheet.html]: Yes, that makes more sense. I updated the code.

**Comment by Manishearth** [comment]
@bors-servo r+

thanks!

**Pull request merged**